### ISSUE TITLE
4.5.1 fails to build on macOS

### ISSUE THREAD

**Issue by lock042** [opencv/opencv#20425]
While I've no issue with previous versions, I can't compile this one on macOS.

Here the error:

```
FAILED: modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/macosx_conversions.mm.o 
/Applications/Xcode-10.1.app/Contents/Developer/usr/bin/g++  -DCVAPI_EXPORTS -DHAVE_IMGCODEC_HDR -DHAVE_IMGCODEC_PFM -DHAVE_IMGCODEC_PXM -DHAVE_IMGCODEC_SUNRASTER -DHAVE_WEBP -DOPJ_STATIC -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/distiller/gtk/source/opencv-4.5.1/3rdparty/openjpeg/openjp2 -I3rdparty/openjpeg/openjp2 -I3rdparty/libtiff -I/Users/distiller/gtk/source/opencv-4.5.1/3rdparty/libtiff -I/Users/distiller/gtk/source/opencv-4.5.1/3rdparty/libpng -I/Users/distiller/gtk/source/opencv-4.5.1/3rdparty/libwebp/src -I3rdparty/libjpeg-turbo -I/Users/distiller/gtk/source/opencv-4.5.1/3rdparty/libjpeg-turbo/src -I3rdparty/zlib -I/Users/distiller/gtk/source/opencv-4.5.1/3rdparty/zlib -I/Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/include -Imodules/imgcodecs -I/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include -I/Users/distiller/gtk/source/opencv-4.5.1/modules/imgproc/include -isystem . -O3 -arch x86_64 -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.12.sdk/usr/include -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.12.sdk -mmacosx-version-min=10.9   -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wno-delete-non-virtual-dtor -Wno-unnamed-type-template-args -Wno-comment -fdiagnostics-show-option -Wno-long-long -Qunused-arguments -Wno-semicolon-before-method-body -ffunction-sections -fdata-sections  -fvisibility=hidden -fvisibility-inlines-hidden -Wno-deprecated-declarations -O3 -DNDEBUG  -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.12.sdk -mmacosx-version-min=10.9 -fPIC   -std=c++11 -MD -MT modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/macosx_conversions.mm.o -MF modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/macosx_conversions.mm.o.d -o modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/macosx_conversions.mm.o -c /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/macosx_conversions.mm
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/macosx_conversions.mm:6:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/apple_conversions.h:8:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core.hpp:54:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/base.hpp:58:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvstd.hpp:81:
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvstd_wrapper.hpp:45:160: error: too many arguments provided to function-like macro invocation
    static CV_CONSTEXPR std::true_type check(typename std::is_same<typename std::decay<decltype(std::declval<T>().operator()(std::declval<Args>()...))>::type, Ret>::type*);
                                                                                                                                                               ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.12.sdk/usr/include/AssertMacros.h:1291:10: note: macro 'check' defined here
        #define check(assertion)  __Check(assertion)
                ^
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/macosx_conversions.mm:6:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/apple_conversions.h:8:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core.hpp:54:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/base.hpp:58:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvstd.hpp:81:
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvstd_wrapper.hpp:45:40: warning: variable templates are a C++14 extension [-Wc++14-extensions]
    static CV_CONSTEXPR std::true_type check(typename std::is_same<typename std::decay<decltype(std::declval<T>().operator()(std::declval<Args>()...))>::type, Ret>::type*);
                                       ^
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvstd_wrapper.hpp:45:40: error: declaration of constexpr static data member 'check' requires an initializer
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvstd_wrapper.hpp:47:31: error: constexpr can only be used in variable and function declarations
    template<typename> static CV_CONSTEXPR std::false_type check(...);
                              ^
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/cvdef.h:757:26: note: expanded from macro 'CV_CONSTEXPR'
#    define CV_CONSTEXPR constexpr
                         ^
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/macosx_conversions.mm:6:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/imgcodecs/src/apple_conversions.h:8:
In file included from /Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core.hpp:3304:
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/utility.hpp:53:4: warning: Detected Apple 'check' macro definition, it can cause build conflicts. Please, include this header before any Apple headers. [-W#warnings]
#  warning Detected Apple 'check' macro definition, it can cause build conflicts. Please, include this header before any Apple headers.
   ^
/Users/distiller/gtk/source/opencv-4.5.1/modules/core/include/opencv2/core/utility.hpp:905:5: warning: declaration does not declare anything [-Wmissing-declarations]
    bool check() const;
    ^~~~~~~~~~~~~~~~~~
3 warnings and 3 errors generated.
[740/955] Building CXX object modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/loadsave.cpp.o
[741/955] Building CXX object modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/grfmt_gdal.cpp.o
[742/955] Building CXX object modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/grfmt_gdcm.cpp.o
[743/955] Building CXX object modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/grfmt_hdr.cpp.o
[744/955] Building CXX object modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/grfmt_jpeg.cpp.o
ninja: build stopped: subcommand failed.
*** Error during phase build of opencv: ########## Error running ninja   *** [12/32]
```

**Comment by alalek** [comment]
No way to amend already released code.
You need to upgrade. See fix here: #19468

---

Build / usage questions should go to Users OpenCV Q/A forum: https://forum.opencv.org/

**Comment by lock042** [comment]
@alalek : thx for the link. I haven't see that. I did not know it was fixed.
Thx.
